meshio/msh_io/msh4_1.py

```diff
@@ -0,0 +1,452 @@
+            fromfile(f, c_double, 3 if d == 0 else 6)  # discard bounding-box
+            num_physicals, = fromfile(f, c_ulong, 1)
+            physical_tags[d][tag] = list(fromfile(f, c_int, num_physicals))
+            if d > 0:  # discard tagBREP{Vert,Curve,Surfaces}
+                num_BREP_, = fromfile(f, c_ulong, 1)
+                fromfile(f, c_int, num_BREP_)
+
+    if not is_ascii:
+        line = f.readline().decode("utf-8").strip()
+        assert line == ""
+    line = f.readline().decode("utf-8").strip()
+    assert line == "$EndEntities"
+    return physical_tags
+
+
+def _read_nodes(f, is_ascii, int_size, data_size):
+    fromfile = partial(numpy.fromfile, sep=" " if is_ascii else "")
+
+    # numEntityBlocks numNodes minNodeTag maxNodeTag (all size_t)
+    num_entity_blocks, total_num_nodes, _, __ = fromfile(f, c_size_t, 4)
+
+    points = numpy.empty((total_num_nodes, 3), dtype=float)
+    tags = numpy.empty(total_num_nodes, dtype=int)
+
+    idx = 0
+    for k in range(num_entity_blocks):
+        # entityDim(int) entityTag(int) parametric(int) numNodes(size_t)
+        _, __, parametric = fromfile(f, c_int, 3)
+        assert parametric == 0, "parametric nodes not implemented"
+        num_nodes = int(fromfile(f, c_size_t, 1)[0])
+
+        node_tags = fromfile(f, c_size_t, num_nodes)
+        xyz = fromfile(f, c_double, num_nodes * 3).reshape((num_nodes, 3))
+        points[idx : (idx + num_nodes)] = xyz
+        tags[idx : (idx + num_nodes)] = node_tags
+        idx += num_nodes
+
+    if not is_ascii:
+
+        line = f.readline().decode("utf-8")
+        assert line == "\n"
+
+    line = f.readline().decode("utf-8")
+    assert line.strip() == "$EndNodes"
+
+    return points, tags
+
+
+def _read_cells(f, point_tags, int_size, is_ascii, physical_tags):
+    fromfile = partial(numpy.fromfile, sep=" " if is_ascii else "")
+
+    # numEntityBlocks numElements minElementTag maxElementTag (all size_t)
+    num_entity_blocks, total_num_elements, _, __ = fromfile(f, c_size_t, 4)
+
+    data = []
+    cell_data = {}
+
+    for k in range(num_entity_blocks):
+        # entityDim(int) entityTag(int) elementType(int) numElements(size_t)
+        dim_entity, tag_entity, type_ele = fromfile(f, c_int, 3)
+        num_ele, = fromfile(f, c_size_t, 1)
+        tpe = _gmsh_to_meshio_type[type_ele]
+        num_nodes_per_ele = num_nodes_per_cell[tpe]
+        d = fromfile(f, c_size_t, int(num_ele * (1 + num_nodes_per_ele))).reshape(
+            (num_ele, -1)
+        )
+        if physical_tags is None:
+            data.append((None, tpe, d))
+        else:
+            data.append((physical_tags[dim_entity][tag_entity], tpe, d))
+
+    if not is_ascii:
+        line = f.readline().decode("utf-8")
+        assert line == "\n"
+    line = f.readline().decode("utf-8")
+    assert line.strip() == "$EndElements"
+
+    # The msh4 elements array refers to the nodes by their tag, not the index. All other
```

Comment:
Here, we could check if the tag are already in order (as pointed out by Christophe), and if yes, set `itags` directly.

Reply:
Yes.  1bb3111555a31f0ef1b47d3dc46b134a55ccda24